File: ds3_java_cli/src/main/java/com/spectralogic/ds3cli/views/cli/TableView.java
/*
 * ***************************************************************************
 *   Copyright 2014-2019 Spectra Logic Corporation. All Rights Reserved.
 *   Licensed under the Apache License, Version 2.0 (the "License"). You may not use
 *   this file except in compliance with the License. A copy of the License is located at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file.
 *   This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 *   CONDITIONS OF ANY KIND, either express or implied. See the License for the
 *   specific language governing permissions and limitations under the License.
 * ***************************************************************************
 */

package com.spectralogic.ds3cli.views.cli;

import com.bethecoder.ascii_table.ASCIITable;
import com.bethecoder.ascii_table.ASCIITableHeader;
import com.google.common.collect.ImmutableList;
import com.spectralogic.ds3cli.View;
import com.spectralogic.ds3cli.models.Result;

public abstract class TableView<T extends Result> implements View<T> {

    protected ASCIITableHeader[] header;
    protected int columnCount;

    public abstract String render(final T result);

    public void initTable(final ImmutableList<String> columnHeads) {
        this.columnCount = columnHeads.size();

        // create the header
        this.header = new ASCIITableHeader[this.columnCount];
        for (int i = 0; i < this.columnCount; i++) {
            header[i] = new ASCIITableHeader(columnHeads.get(i), ASCIITable.ALIGN_LEFT);
        }
    }

    public void setTableDataAlignment(final ImmutableList<Integer> columnAlign) {
        // set alignment
        if (this.header != null && this.columnCount > 0) {
            for (int i = 0; i < this.columnCount; i++) {
                this.header[i].setDataAlign(columnAlign.get(i).shortValue());
            }
        }
    }

    protected String renderTable() {
        return ASCIITable.getInstance().getTable(getHeaders(), formatTableContents());
    }

    protected abstract String[][] formatTableContents();

    protected ASCIITableHeader[] getHeaders() {
        return this.header;
    }
}

File: ds3-cli-tableview/src/main/kotlin/com/spectralogic/ds3cli/tableview/TableView.kt
/*
 * *****************************************************************************
 *   Copyright 2014-2016 Spectra Logic Corporation. All Rights Reserved.
 *   Licensed under the Apache License, Version 2.0 (the "License"). You may not use
 *   this file except in compliance with the License. A copy of the License is located at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file.
 *   This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 *   CONDITIONS OF ANY KIND, either express or implied. See the License for the
 *   specific language governing permissions and limitations under the License.
 * ***************************************************************************
 */

package com.spectralogic.ds3cli.tableview

import com.bethecoder.ascii_table.ASCIITable
import com.spectralogic.ds3cli.utils.Constants.DATE_FORMAT
import com.spectralogic.ds3cli.utils.Guard.*
import com.spectralogic.ds3cli.tableview.table.AsciiTable
import java.util.*

interface TableView {
    fun print(): String
}

class AsciiTableView<in T>(columns: List<TableEntry<T>>,
                           data: Iterable<T>) : TableView {

    private val backingTable: AsciiTable<T>

    init {
        backingTable = AsciiTable(columns, data)
    }

    override fun print(): String {

        return ASCIITable.getInstance().getTable(backingTable) ?: ""
    }
}

class CsvTableView<in T>(columns: List<TableEntry<T>>, data: Iterable<T>) : TableView {
    override fun print(): String {
        throw NotImplementedError("This has not been implemented yet")
    }
}

enum class TableEntryAlignment {
    LEFT, RIGHT
}

interface TableEntry<in T> {
    val alignment: TableEntryAlignment
    val columnName: String
    fun formatCell(entry: T): String
}

abstract class TableLeftAlignedEntry<in T> : TableEntry<T> {
    override val alignment: TableEntryAlignment = TableEntryAlignment.LEFT
}

class TableIdEntry<in T>(override val columnName: String,
                         val transform: (T) -> UUID?) : TableLeftAlignedEntry<T>() {

    override fun formatCell(entry: T): String {
        return nullGuardToString(transform.invoke(entry))
    }
}

class TableStringEntry<in T>(override val columnName: String,
                             val transform: (T) -> String?) : TableLeftAlignedEntry<T>() {

    override fun formatCell(entry: T): String {
        return nullGuard(transform.invoke(entry))
    }
}

class TableBooleanEntry<in T>(override val columnName: String,
                           val transform: (T) -> Boolean?) : TableLeftAlignedEntry<T>() {

    override fun formatCell(entry: T): String {
        return nullGuardToString(transform.invoke(entry))
    }
}

class TableLongEntry<in T>(override val columnName: String,
                           val transform: (T) -> Long?) : TableLeftAlignedEntry<T>() {

    override fun formatCell(entry: T): String {
        return nullGuardToString(transform.invoke(entry))
    }
}

class TableDateEntry<in T>(override val columnName: String,
                           val transform: (T) -> Date?) : TableLeftAlignedEntry<T>() {

    override fun formatCell(entry: T): String {
        return nullGuardFromDate(transform.invoke(entry), DATE_FORMAT)
    }
}
